### PR TITLE
Linus database synchronisation

### LINKED ISSUE #108: Transfer between (Appointment and LinusAppointment) and (ExperimentBooking and LinusExperimentBooking)
LinusAppointments have to be transformed into Appointments. LinusAppointments are static, Appointments are not. All LinusAppointments that not exist in Appointments, have to be imported. Appointments can be moved or deleted. When this happend, the corresponding LinusAppointment must nut be imported again.

Same for ExperimentBooking and LinusExperimentBooking.

### PR DESCRIPTION
**Pull request by MarcoLeipnitz**
This PR introduces an possibility for us to convert entities from the read-only linus db to entities that we can manage.

closes #108

---

### Implementation

- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified without unreasonable effort
- [ ] Is the code modular enough?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [ ] Interface Segregation: If interface implemented, is it fully implemented?
- [ ] Dependency Injection: Are central object instantiation processes managed by a single context?
- [ ] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [ ] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [ ] Can you think of any use case in which the code does not behave as intended?
- [ ] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information be added or removed?
- [ ] Are error messages user-friendly?

### Dependencies (internal and external)

- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Is the API.md up-to-date?
- [ ] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [ ] Does the code introduce any security vulnerabilities?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is sensitive data like user data handled and stored securely?
- [ ] Is the right encryption used?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [ ] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [ ] Is the API/UI intuitive to use?
- [ ] Was the app tested on another device?

### Readability

- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by smaller methods/different names?
- [ ] Is the code located in the right file/folder?
- [ ] Do you think certain methods should be restructured to have a more intuitive control flow?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message better?
- [ ] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/User.java [hunk 1]
```diff
@@ -30,28 +35,14 @@ public class User {
...
-
-    public void setEmail(String email) {
-        this.email = email;
-    }
+    /**
+     * User Id in linus (if existent)
+     */
+    @Nullable
+    @Column(name = "linus_user_id")
+    private Integer linusUserId;
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/User.java]: For what do we need this? We have such an attribute at Person already.

**Review by JohaJung — COMMENTED** [review]
Habe mir den interessanten Teil noch nicht angeguckt, aber hier konzeptionelle Fragen

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            ExperimentBooking newEntry = new ExperimentBooking();
+            newEntry.setLinusExperimentBookingId(toBeSavedBooking.getId());
+            newEntry.setLinusExperimentId(toBeSavedBooking.getLinusExperimentId());
+            List<Person> person = vlaPersonDb.findByLinusUserId(toBeSavedBooking.getLinusUserId());
+            // TODO: make new person if there is no person with linusUserId (linusUser needed)
+            if (!person.isEmpty()) {
+                newEntry.setPerson(person.getFirst());
+            }
+            // TODO: appointment
+            // no notes here because in linus notes are attached to the appointment
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java]: I propose to add the notes of the LinusAppointment to every LinusExperimentBooking

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        linusExperiments.stream().filter(linusExperimentBooking ->
+        {
+            return experimentBookingRepository
+                .findByLinusExperimentBookingId(linusExperimentBooking.getId()).isEmpty();
+        }).forEach(toBeSavedBooking -> {
+            ExperimentBooking newEntry = new ExperimentBooking();
+            newEntry.setLinusExperimentBookingId(toBeSavedBooking.getId());
+            newEntry.setLinusExperimentId(toBeSavedBooking.getLinusExperimentId());
+            List<Person> person = vlaPersonDb.findByLinusUserId(toBeSavedBooking.getLinusUserId());
+            // TODO: make new person if there is no person with linusUserId (linusUser needed)
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java]: Hmm, unsure whether this auto-generation of Person entities in the backend is a good idea.

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    @Transactional
+    public void syncAppointments(LocalDateTime start, LocalDateTime end) {
+        List<LinusAppointment> linusAppointments =
+            linusAppointmentRepository.findByAppointmentTimeBetween(start, end);
+
+        List<Appointment> newAppointments = new ArrayList<>();
+
+        Optional<AppointmentCategory> fetchedCategory =
+            appointmentCategoryRepository.findByTitle("LinusAppointmentImports");
+
```
Comment (JohaJung) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java]: Urgh. This is very ugly special casing

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java (hunk 3 above)
Comment (MarcoLeipnitz) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java]: I am also uncertain, this is a note @CarolinRoell wrote down.
I think it would be better just to append the linus Name to the notes of experimentBooking or somewhere else.

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java [hunk 5]
```diff
@@ -48,5 +59,13 @@ public class Appointment {
     @Column(name = "notes", nullable = false)
     private String notes = "";
 
+    /**
+     * Corresponding linus appointment id (if existent).
+     */
+    @JsonIgnore
+    @Nullable
+    @Column(name = "linus_appointment_id")
+    private Integer linusAppointmentId;
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/vladb/Appointment.java]: > When we look at your matching idea below, I think we can get rid of this entry

Exactly :100:

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java (hunk 4 above)
Comment (JohaJung) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/databaseSync/LinusSyncService.java]: > What would you propose otherwise?

Do not automatically create any appointment at all :wink:, see below conversation. This also removes the need for creating an appointment category.

**Comment by MarcoLeipnitz** [comment]
Functionality-wise this should be complete in the backend @JohaJung 
I will clean this up and then we should be ready to go.

Should the frontend also be part of this PR?
This will probably inflate the things to review even more.

**Review by JohaJung — CHANGES_REQUESTED** [review]

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java [hunk 6]
```diff
@@ -85,15 +108,22 @@ public ResponseEntity<?> updateAppointment(@PathVariable Long id,
...
-        Appointment deletedAppointment = appointmentRepository.findById(id).orElseThrow(
+        Appointment toBeDeletedAppointment = appointmentRepository.findById(id).orElseThrow(
             () -> new EntityNotFoundException(
                 "Appointment with ID " + id + " not found."));
+
+        if (!toBeDeletedAppointment.getBookings().isEmpty()) {
+            throw new InvalidParameterException(
+                "Appointments with assigned experimentBookings may not be deleted"
+            );
+        }
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java]: Let us further discuss this in #122 as that PR is currently significantly smaller and tailored to this method

**Review by CarolinRoell — CHANGES_REQUESTED** [review]

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java]: This part is now ready for review in #122

File: VLA-Backend/src/test/java/de/vlaorgatu/vlabackend/TestUtils.java [hunk 7]
```diff
@@ -1,10 +1,27 @@
 package de.vlaorgatu.vlabackend;
 
+<<<<<<< HEAD
```
Comment (JohaJung) [inline VLA-Backend/src/test/java/de/vlaorgatu/vlabackend/TestUtils.java]: Please take a look at the merge conflicts in this file, I did not dig deep enough into the logic to understand what to choose.
